dev/com.ibm.ws.jdbc/src/com/ibm/ws/jdbc/internal/PropertyService.java

```diff
@@ -152,9 +153,131 @@ public int hashCode() {
+            if(beginToken != progress) {
+                sb.append("****" + url.charAt(beginToken));
+            } 
+            
+            String propertyName = url.substring(beginToken + 1, equals + 1);
+            sb.append(propertyName);
+            
+            int endToken;
+            if(lowerCaseURL.charAt(beginToken) == '(')
+                endToken = url.indexOf(')', equals + 1);
+            else
+                endToken = url.indexOf(lowerCaseURL.charAt(beginToken), equals + 1);
+
+            String propertyValue = "";
+            
+            if(endToken == -1) {
+                //check to see if it is because we are on the last property
+                if(url.indexOf("=", equals + 1) == -1) {
+                    //we've reached the end, append the property value
+                    endToken = url.length() - 1;
+                } else {
+                    //we aren't at the end, but we haven't come across the beginning token either
+                    Character beginTokenChar = url.charAt(beginToken);
+                    //try some common possibilities for mismatched beginning/ending token
+                    switch(beginTokenChar) {
+                        case '?' :  
+                            endToken = url.indexOf("&", equals);
+                            break;
+                        case ':' :
+                            endToken = url.indexOf(";", equals);
+                            break;
+                        case ',' :
+                            endToken = url.indexOf(")", equals);
+                            break;
+                        case '(' :
+                            endToken = url.indexOf(",", equals);
+                            break;
+                            //no default as it is handled below
+                    }
+                    if(endToken == -1) {
+                        //we still haven't found the right combination of beginning and ending tokens,
+                        //so we will just star out this property and move on
+                        progress = url.indexOf("=", equals + 1) - 1; //set progress to right before the next =
+                        sb.append("****");
+                        continue;
+                    } else {
+                        propertyValue = Character.toString(url.charAt(endToken));
+                    }
+                }
+            } else {
+                propertyValue = Character.toString(url.charAt(endToken));
+            }
+            
+            if(endToken == equals) {
+                //nothing to append
+            } else if(endToken == equals + 1 && url.charAt(endToken) == url.charAt(beginToken)) {
+                sb.append(propertyValue);
+            }  else if(propertyName.toLowerCase().contains("password") || propertyName.toLowerCase().contains("pwd")) {
```

Comment:
Are there really observed cases of JDBC drivers naming password properties "pwd"?  Do we need to check for that?

Reply:
Yes, there are JDBC drivers using PWD